test/dt_arith.c

```diff
@@ -4043,829 +4479,3588 @@ test_conv_int_fp(const char *name, int run_test, hid_t src, hid_t dst)
+            printf(" %29.20Le%+.20Lei\n", creall(hw_ldouble_complex), cimagl(hw_ldouble_complex));
+#endif
+        else
+            goto error;
+
+        /* If the source is normalized values, print out error message; if it is
+         * denormalized or special values, print out warning message.*/
+        if (++fails_all_tests >= max_fails) {
+            if (run_test == TEST_NORMAL)
+                puts("    maximum failures reached, aborting test...");
+            else if (run_test == TEST_DENORM || run_test == TEST_SPECIAL)
+                puts("    maximum warnings reached, aborting test...");
+            puts("    (dst is library's conversion output. ans is compiler's conversion output.)");
+
+            goto done;
+        }
+    }
+
+    if (!fails_all_tests)
+        PASSED();
+
+done:
+    if (buf)
+        aligned_free(buf);
+    if (saved)
+        aligned_free(saved);
+    fflush(stdout);
+#ifdef HANDLE_SIGFPE
+    if (run_test == TEST_NOOP || run_test == TEST_NORMAL)
+        exit(MIN((int)fails_all_tests, 254));
+    else if (run_test == TEST_DENORM || run_test == TEST_SPECIAL)
+        exit(EXIT_SUCCESS);
+    assert(0 && "Should not reach this point!");
+    return 1;
+#else
+    /* Restore the default error handler (set in h5_test_init()) */
+    h5_restore_err();
+
+    reset_hdf5();
+
+    /* If the source is normalized values, treat the failures as error;
+     * if it is denormalized or special values, treat the failure as warning.*/
+    if (run_test == TEST_NOOP || run_test == TEST_NORMAL)
+        return (int)fails_all_tests;
+    else if (run_test == TEST_DENORM || run_test == TEST_SPECIAL)
+        return 0;
+#endif
+
+error:
+    if (buf)
+        aligned_free(buf);
+    if (saved)
+        aligned_free(saved);
+    fflush(stdout);
+#ifdef HANDLE_SIGFPE
+    if (run_test == TEST_NOOP || run_test == TEST_NORMAL)
+        exit(MIN(MAX((int)fails_all_tests, 1), 254));
+    else if (run_test == TEST_DENORM || run_test == TEST_SPECIAL)
+        exit(EXIT_FAILURE);
+    assert(0 && "Should not reach this point!");
+    return 1;
+#else
+    /* Restore the default error handler (set in h5_test_init()) */
+    h5_restore_err();
+
+    reset_hdf5();
+
+    if (run_test == TEST_NOOP || run_test == TEST_NORMAL)
+        return MAX((int)fails_all_tests, 1);
+    else if (run_test == TEST_DENORM || run_test == TEST_SPECIAL)
+        return 1;
+    return 1;
+#endif
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    test_conv_int_fp_conv_to_schar
+ *
+ * Purpose:     Helper function for test_conv_int_fp to perform conversion
+ *              from a datatype to signed char by casting.
+ *
+ * Return:      -1 on failure
+ *              0 on success
+ *
+ *-------------------------------------------------------------------------
+ */
+static int
+test_conv_int_fp_conv_to_schar(void *hw_p, unsigned char *src_buf, size_t idx, dtype_t src_type)
```

Comment:
Also similar to the changes in the `_Float16` PR, the logic in `test_conv_int_fp` was getting a bit lengthy due to all the types supported, including the new complex number types. I moved each type into a helper function and `test_conv_int_fp` moved below these helper functions. It make this particular diff difficult to follow, but makes the test much easier to read and follow.